File: apps/c/simulator/quantum.c
/*
 * quantum.c - Quantum field simulation implementation.
 *
 * Models quantum fields, particles, wave functions, pair production,
 * quark-hadron transition. Faithfully ports the Python simulator logic.
 */
#include "quantum.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>

/* ------------------------------------------------------------------ */
/* Helpers                                                             */
/* ------------------------------------------------------------------ */

static int g_particle_id_counter = 0;

static float randf(void)
{
    return (float)rand() / (float)RAND_MAX;
}

/* Box-Muller transform for Gaussian random numbers */
static float gaussf(float mean, float stddev)
{
    float u1 = randf();
    float u2 = randf();
    if (u1 < 1e-10f) u1 = 1e-10f;
    float z = sqrtf(-2.0f * logf(u1)) * cosf(2.0f * (float)SIM_PI * u2);
    return mean + stddev * z;
}

/* Exponential variate with rate lambda */
static float expovariatef(float lambda)
{
    float u = randf();
    if (u < 1e-10f) u = 1e-10f;
    return -logf(u) / lambda;
}

/* ------------------------------------------------------------------ */
/* Particle helpers                                                    */
/* ------------------------------------------------------------------ */

static const float MASS_TABLE[PTYPE_COUNT] = {
    [PTYPE_UP]       = M_UP_QUARK,
    [PTYPE_DOWN]     = M_DOWN_QUARK,
    [PTYPE_ELECTRON] = M_ELECTRON,
    [PTYPE_POSITRON] = M_ELECTRON,
    [PTYPE_NEUTRINO] = (float)M_NEUTRINO,
    [PTYPE_PHOTON]   = M_PHOTON,
    [PTYPE_GLUON]    = M_PHOTON,
    [PTYPE_W_BOSON]  = M_W_BOSON,
    [PTYPE_Z_BOSON]  = M_Z_BOSON,
    [PTYPE_PROTON]   = M_PROTON,
    [PTYPE_NEUTRON]  = M_NEUTRON,
};

static const float CHARGE_TABLE[PTYPE_COUNT] = {
    [PTYPE_UP]       =  2.0f / 3.0f,
    [PTYPE_DOWN]     = -1.0f / 3.0f,
    [PTYPE_ELECTRON] = -1.0f,
    [PTYPE_POSITRON] =  1.0f,
    [PTYPE_NEUTRINO] =  0.0f,
    [PTYPE_PHOTON]   =  0.0f,
    [PTYPE_GLUON]    =  0.0f,
    [PTYPE_W_BOSON]  =  1.0f,
    [PTYPE_Z_BOSON]  =  0.0f,
    [PTYPE_PROTON]   =  1.0f,
    [PTYPE_NEUTRON]  =  0.0f,
};

static const char *TYPE_NAMES[PTYPE_COUNT] = {
    "up", "down", "electron", "positron", "neutrino",
    "photon", "gluon", "W", "Z", "proton", "neutron"
};

float particle_mass(ParticleType t)
{
    if (t >= 0 && t < PTYPE_COUNT) return MASS_TABLE[t];
    return 0.0f;
}

float particle_charge(ParticleType t)
{
    if (t >= 0 && t < PTYPE_COUNT) return CHARGE_TABLE[t];
    return 0.0f;
}

float particle_energy(const Particle *p)
{
    float p2 = p->momentum[0] * p->momentum[0]
             + p->momentum[1] * p->momentum[1]
             + p->momentum[2] * p->momentum[2];
    float m  = particle_mass(p->type);
    float c  = (float)SIM_C;
    return sqrtf(p2 * c * c + m * m * c * c * c * c);
}

const char *particle_type_name(ParticleType t)
{
    if (t >= 0 && t < PTYPE_COUNT) return TYPE_NAMES[t];
    return "unknown";
}

/* ------------------------------------------------------------------ */
/* Wave function                                                       */
/* ------------------------------------------------------------------ */

void wf_init(WaveFunction *wf)
{
    wf->amplitude = 1.0f;
    wf->phase     = 0.0f;
    wf->coherent  = true;
}

void wf_evolve(WaveFunction *wf, float dt, float energy)
{
    if (wf->coherent) {
        wf->phase += energy * dt / (float)SIM_HBAR;
        wf->phase = fmodf(wf->phase, 2.0f * (float)SIM_PI);
    }
}

bool wf_collapse(WaveFunction *wf)
{
    float prob = wf->amplitude * wf->amplitude;
    bool result = randf() < prob;
    wf->amplitude = result ? 1.0f : 0.0f;
    wf->coherent  = false;
    return result;
}

/* ------------------------------------------------------------------ */
/* Quantum field                                                       */
/* ------------------------------------------------------------------ */

void qf_init(QuantumField *qf, double temperature)
{
    qf->capacity          = 256;
    qf->count             = 0;
    qf->particles         = (Particle *)calloc((size_t)qf->capacity, sizeof(Particle));
    qf->temperature       = temperature;
    qf->vacuum_energy     = 0.0;
    qf->total_created     = 0;
    qf->total_annihilated = 0;
}

void qf_free(QuantumField *qf)
{
    free(qf->particles);
    qf->particles = NULL;
    qf->count = 0;
    qf->capacity = 0;
}

static void qf_ensure_capacity(QuantumField *qf, int needed)
{
    if (qf->count + needed <= qf->capacity) return;
    int new_cap = qf->capacity * 2;
    while (new_cap < qf->count + needed) new_cap *= 2;
    if (new_cap > MAX_PARTICLES) new_cap = MAX_PARTICLES;
    qf->particles = (Particle *)realloc(qf->particles,
                                        (size_t)new_cap * sizeof(Particle));
    qf->capacity = new_cap;
}

int qf_add_particle(QuantumField *qf, const Particle *p)
{
    if (qf->count >= MAX_PARTICLES) return -1;
    qf_ensure_capacity(qf, 1);
    int idx = qf->count;
    qf->particles[idx] = *p;
    qf->particles[idx].particle_id = ++g_particle_id_counter;
    qf->count++;
    return idx;
}

void qf_remove_particle(QuantumField *qf, int index)
{
    if (index < 0 || index >= qf->count) return;
    /* Swap with last element for O(1) removal */
    qf->particles[index] = qf->particles[qf->count - 1];
    qf->count--;
}

void qf_evolve(QuantumField *qf, float dt)
{
    for (int i = 0; i < qf->count; i++) {
        Particle *p = &qf->particles[i];
        float m = particle_mass(p->type);

        if (m > 0.0f) {
            /* Massive particles: dp/m * dt */
            for (int j = 0; j < 3; j++) {
                p->position[j] += p->momentum[j] / m * dt;
            }
        } else {
            /* Massless: move at c */
            float p_mag = sqrtf(p->momentum[0] * p->momentum[0]
                              + p->momentum[1] * p->momentum[1]
                              + p->momentum[2] * p->momentum[2]);
            if (p_mag < 1e-20f) p_mag = 1.0f;
            for (int j = 0; j < 3; j++) {
                p->position[j] += p->momentum[j] / p_mag * (float)SIM_C * dt;
            }
        }

        /* Evolve wave function */
        float e = particle_energy(p);
        wf_evolve(&p->wave_fn, dt, e);
    }
}

bool qf_pair_production(QuantumField *qf, float energy)
{
    if (energy < 2.0f * M_ELECTRON * (float)(SIM_C * SIM_C))
        return false;
    if (qf->count + 2 > MAX_PARTICLES)
        return false;

    ParticleType p_type, ap_type;
    if (energy >= 2.0f * M_PROTON * (float)(SIM_C * SIM_C) && randf() < 0.1f) {
        p_type  = PTYPE_UP;
        ap_type = PTYPE_DOWN;
    } else {
        p_type  = PTYPE_ELECTRON;
        ap_type = PTYPE_POSITRON;
    }

    float dir[3];
    dir[0] = gaussf(0.0f, 1.0f);
    dir[1] = gaussf(0.0f, 1.0f);
    dir[2] = gaussf(0.0f, 1.0f);
    float norm = sqrtf(dir[0]*dir[0] + dir[1]*dir[1] + dir[2]*dir[2]);
    if (norm < 1e-10f) norm = 1.0f;
    float p_mom = energy / (2.0f * (float)SIM_C);

    Particle pa = {0}, pb = {0};
    pa.type = p_type;
    pb.type = ap_type;
    pa.spin = SPIN_UP;
    pb.spin = SPIN_DOWN;
    wf_init(&pa.wave_fn);
    wf_init(&pb.wave_fn);
    pa.entangled_with = -1;
    pb.entangled_with = -1;

    for (int i = 0; i < 3; i++) {
        pa.momentum[i] =  (dir[i] / norm) * p_mom;
        pb.momentum[i] = -(dir[i] / norm) * p_mom;
        pa.position[i] = 0.0f;
        pb.position[i] = 0.0f;
    }

    int ia = qf_add_particle(qf, &pa);
    int ib = qf_add_particle(qf, &pb);
    if (ia >= 0 && ib >= 0) {
        qf->particles[ia].entangled_with = qf->particles[ib].particle_id;
        qf->particles[ib].entangled_with = qf->particles[ia].particle_id;
    }
    qf->total_created += 2;
    return true;
}

bool qf_vacuum_fluctuation(QuantumField *qf)
{
    double prob = qf->temperature / T_PLANCK;
    if (prob > 0.5) prob = 0.5;
    if (randf() < (float)prob) {
        float inv_lambda = (float)(qf->temperature * 0.001);
        if (inv_lambda < 1e-10f) inv_lambda = 1e-10f;
        float energy = expovariatef(1.0f / inv_lambda);
        return qf_pair_production(qf, energy);
    }
    return false;
}

int qf_quark_confinement(QuantumField *qf)
{
    if (qf->temperature > T_QUARK_HADRON) return 0;

    int hadrons_formed = 0;

    /* Collect indices of ups and downs */
    int ups[MAX_PARTICLES], n_ups = 0;
    int downs[MAX_PARTICLES], n_downs = 0;

    for (int i = 0; i < qf->count; i++) {
        if (qf->particles[i].type == PTYPE_UP   && n_ups   < MAX_PARTICLES)
            ups[n_ups++] = i;
        if (qf->particles[i].type == PTYPE_DOWN && n_downs < MAX_PARTICLES)
            downs[n_downs++] = i;
    }

    /* Form protons: uud */
    while (n_ups >= 2 && n_downs >= 1) {
        int iu1 = ups[--n_ups];
        int iu2 = ups[--n_ups];
        int id1 = downs[--n_downs];

        Particle proton = {0};
        proton.type = PTYPE_PROTON;
        proton.spin = SPIN_UP;
        proton.entangled_with = -1;
        wf_init(&proton.wave_fn);

        for (int j = 0; j < 3; j++) {
            proton.position[j] = qf->particles[iu1].position[j];
            proton.momentum[j] = qf->particles[iu1].momentum[j]
                               + qf->particles[iu2].momentum[j]
                               + qf->particles[id1].momentum[j];
        }

        /* Mark quarks for removal (set type to count=invalid, handled below) */
        qf->particles[iu1].type = PTYPE_COUNT;
        qf->particles[iu2].type = PTYPE_COUNT;
        qf->particles[id1].type = PTYPE_COUNT;

        qf_add_particle(qf, &proton);
        hadrons_formed++;
    }

    /* Form neutrons: udd */
    while (n_ups >= 1 && n_downs >= 2) {
        int iu1 = ups[--n_ups];
        int id1 = downs[--n_downs];
        int id2 = downs[--n_downs];

        Particle neutron = {0};
        neutron.type = PTYPE_NEUTRON;
        neutron.spin = SPIN_UP;
        neutron.entangled_with = -1;
        wf_init(&neutron.wave_fn);

        for (int j = 0; j < 3; j++) {
            neutron.position[j] = qf->particles[iu1].position[j];
            neutron.momentum[j] = qf->particles[iu1].momentum[j]
                                + qf->particles[id1].momentum[j]
                                + qf->particles[id2].momentum[j];
        }

        qf->particles[iu1].type = PTYPE_COUNT;
        qf->particles[id1].type = PTYPE_COUNT;
        qf->particles[id2].type = PTYPE_COUNT;

        qf_add_particle(qf, &neutron);
        hadrons_formed++;
    }

    /* Compact: remove all particles marked as PTYPE_COUNT */
    int write = 0;
    for (int read = 0; read < qf->count; read++) {
        if (qf->particles[read].type != PTYPE_COUNT) {
            if (write != read)
                qf->particles[write] = qf->particles[read];
            write++;
        }
    }
    qf->count = write;

    return hadrons_formed;
}

int qf_particle_count(const QuantumField *qf, ParticleType type)
{
    int count = 0;
    for (int i = 0; i < qf->count; i++) {
        if (qf->particles[i].type == type) count++;
    }
    return count;
}

float qf_total_energy(const QuantumField *qf)
{
    float total = (float)qf->vacuum_energy;
    for (int i = 0; i < qf->count; i++) {
        total += particle_energy(&qf->particles[i]);
    }
    return total;
}

File: apps/wasm/src/quantum.rs
//! Quantum and subatomic physics simulation.
//!
//! Models quantum fields, particles, pair production, annihilation,
//! and the quark-hadron transition. Lightweight port of the Python
//! `simulator/quantum.py` focused on tracking particle counts and
//! positions for rendering.

use rand::Rng;

use crate::constants::*;

// ---------------------------------------------------------------------------
// Particle types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ParticleType {
    Up,
    Down,
    Electron,
    Positron,
    Neutrino,
    Photon,
    Gluon,
    WBoson,
    ZBoson,
    Proton,
    Neutron,
}

impl ParticleType {
    pub fn mass(self) -> f64 {
        match self {
            Self::Up => M_UP_QUARK,
            Self::Down => M_DOWN_QUARK,
            Self::Electron | Self::Positron => M_ELECTRON,
            Self::Neutrino => M_NEUTRINO,
            Self::Photon | Self::Gluon => M_PHOTON,
            Self::WBoson => M_W_BOSON,
            Self::ZBoson => M_Z_BOSON,
            Self::Proton => M_PROTON,
            Self::Neutron => M_NEUTRON,
        }
    }

    pub fn charge(self) -> f64 {
        match self {
            Self::Up => 2.0 / 3.0,
            Self::Down => -1.0 / 3.0,
            Self::Electron => -1.0,
            Self::Positron => 1.0,
            Self::Proton => 1.0,
            _ => 0.0,
        }
    }

    /// Display label for HUD.
    pub fn label(self) -> &'static str {
        match self {
            Self::Up => "up",
            Self::Down => "down",
            Self::Electron => "e-",
            Self::Positron => "e+",
            Self::Neutrino => "nu",
            Self::Photon => "gamma",
            Self::Gluon => "gluon",
            Self::WBoson => "W",
            Self::ZBoson => "Z",
            Self::Proton => "p+",
            Self::Neutron => "n0",
        }
    }

    /// Color for rendering: [r, g, b, a]
    pub fn render_color(self) -> [f32; 4] {
        match self {
            Self::Up => [1.0, 0.2, 0.2, 1.0],       // red
            Self::Down => [0.2, 1.0, 0.2, 1.0],      // green
            Self::Electron => [1.0, 1.0, 0.2, 1.0],  // yellow
            Self::Positron => [1.0, 0.6, 0.2, 1.0],  // orange-ish
            Self::Neutrino => [0.6, 0.6, 1.0, 0.4],  // faint blue
            Self::Photon => [1.0, 1.0, 1.0, 0.9],    // white
            Self::Gluon => [0.3, 0.3, 1.0, 0.6],     // blue
            Self::WBoson => [0.8, 0.3, 0.8, 0.6],    // purple
            Self::ZBoson => [0.3, 0.8, 0.8, 0.6],    // cyan
            Self::Proton => [1.0, 0.55, 0.0, 1.0],   // orange
            Self::Neutron => [0.6, 0.6, 0.6, 1.0],   // gray
        }
    }

    /// Point size for rendering.
    pub fn render_size(self) -> f32 {
        match self {
            Self::Proton | Self::Neutron => 6.0,
            Self::Electron | Self::Positron => 3.0,
            Self::Up | Self::Down => 2.5,
            Self::Photon => 2.0,
            Self::Neutrino => 1.5,
            _ => 2.0,
        }
    }
}

// ---------------------------------------------------------------------------
// Color charge (for quarks)
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ColorCharge {
    Red,
    Green,
    Blue,
}

// ---------------------------------------------------------------------------
// Spin
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Spin {
    Up,
    Down,
}

// ---------------------------------------------------------------------------
// Particle
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct Particle {
    pub id: u32,
    pub particle_type: ParticleType,
    pub position: [f64; 3],
    pub momentum: [f64; 3],
    pub spin: Spin,
    pub color: Option<ColorCharge>,
}

impl Particle {
    pub fn energy(&self) -> f64 {
        let p2: f64 = self.momentum.iter().map(|p| p * p).sum();
        let m = self.particle_type.mass();
        (p2 * C * C + (m * C * C).powi(2)).sqrt()
    }
}

// ---------------------------------------------------------------------------
// Quantum field
// ---------------------------------------------------------------------------

pub struct QuantumField {
    pub temperature: f64,
    pub particles: Vec<Particle>,
    pub vacuum_energy: f64,
    pub total_created: u64,
    pub total_annihilated: u64,
    next_id: u32,
}

impl QuantumField {
    pub fn new(temperature: f64) -> Self {
        Self {
            temperature,
            particles: Vec::new(),
            vacuum_energy: 0.0,
            total_created: 0,
            total_annihilated: 0,
            next_id: 1,
        }
    }

    fn alloc_id(&mut self) -> u32 {
        let id = self.next_id;
        self.next_id = self.next_id.wrapping_add(1);
        id
    }

    /// Seed the early universe with quarks and leptons.
    pub fn seed_early_universe(&mut self, rng: &mut impl Rng) {
        // 30 up quarks
        for _ in 0..30 {
            let id = self.alloc_id();
            self.particles.push(Particle {
                id,
                particle_type: ParticleType::Up,
                position: [gauss(rng, 0.0, 1.0), gauss(rng, 0.0, 1.0), gauss(rng, 0.0, 1.0)],
                momentum: [gauss(rng, 0.0, 5.0), gauss(rng, 0.0, 5.0), gauss(rng, 0.0, 5.0)],
                spin: random_spin(rng),
                color: Some(random_color(rng)),
            });
        }
        // 20 down quarks
        for _ in 0..20 {
            let id = self.alloc_id();
            self.particles.push(Particle {
                id,
                particle_type: ParticleType::Down,
                position: [gauss(rng, 0.0, 1.0), gauss(rng, 0.0, 1.0), gauss(rng, 0.0, 1.0)],
                momentum: [gauss(rng, 0.0, 5.0), gauss(rng, 0.0, 5.0), gauss(rng, 0.0, 5.0)],
                spin: random_spin(rng),
                color: Some(random_color(rng)),
            });
        }
        // 40 electrons
        for _ in 0..40 {
            let id = self.alloc_id();
            self.particles.push(Particle {
                id,
                particle_type: ParticleType::Electron,
                position: [gauss(rng, 0.0, 2.0), gauss(rng, 0.0, 2.0), gauss(rng, 0.0, 2.0)],
                momentum: [gauss(rng, 0.0, 3.0), gauss(rng, 0.0, 3.0), gauss(rng, 0.0, 3.0)],
                spin: random_spin(rng),
                color: None,
            });
        }
        // 5 photons
        for _ in 0..5 {
            let id = self.alloc_id();
            self.particles.push(Particle {
                id,
                particle_type: ParticleType::Photon,
                position: [0.0, 0.0, 0.0],
                momentum: [gauss(rng, 0.0, 10.0), gauss(rng, 0.0, 10.0), gauss(rng, 0.0, 10.0)],
                spin: Spin::Up,
                color: None,
            });
        }
        self.total_created += self.particles.len() as u64;
    }

    /// Pair production from vacuum energy.
    pub fn pair_production(&mut self, energy: f64, rng: &mut impl Rng) -> bool {
        if energy < 2.0 * M_ELECTRON * C * C {
            return false;
        }

        let (p_type, ap_type) = if energy >= 2.0 * M_PROTON * C * C && rng.gen::<f64>() < 0.1 {
            (ParticleType::Up, ParticleType::Down)
        } else {
            (ParticleType::Electron, ParticleType::Positron)
        };

        let dir = [gauss(rng, 0.0, 1.0), gauss(rng, 0.0, 1.0), gauss(rng, 0.0, 1.0)];
        let norm = (dir[0] * dir[0] + dir[1] * dir[1] + dir[2] * dir[2]).sqrt().max(1e-10);
        let p_mag = energy / (2.0 * C);

        let id1 = self.alloc_id();
        let id2 = self.alloc_id();

        self.particles.push(Particle {
            id: id1,
            particle_type: p_type,
            position: [0.0, 0.0, 0.0],
            momentum: [dir[0] / norm * p_mag, dir[1] / norm * p_mag, dir[2] / norm * p_mag],
            spin: Spin::Up,
            color: if p_type == ParticleType::Up {
                Some(random_color(rng))
            } else {
                None
            },
        });
        self.particles.push(Particle {
            id: id2,
            particle_type: ap_type,
            position: [0.0, 0.0, 0.0],
            momentum: [
                -dir[0] / norm * p_mag,
                -dir[1] / norm * p_mag,
                -dir[2] / norm * p_mag,
            ],
            spin: Spin::Down,
            color: if ap_type == ParticleType::Down {
                Some(random_color(rng))
            } else {
                None
            },
        });

        self.total_created += 2;
        true
    }

    /// Vacuum fluctuation: spontaneous pair from vacuum energy.
    pub fn vacuum_fluctuation(&mut self, rng: &mut impl Rng) -> bool {
        let prob = (self.temperature / T_PLANCK).min(0.5);
        if rng.gen::<f64>() < prob {
            // Exponential distribution with mean = temperature * 0.001
            let lambda = 1.0 / (self.temperature * 0.001).max(1e-20);
            let energy = -lambda.recip() * (1.0 - rng.gen::<f64>()).ln();
            return self.pair_production(energy, rng);
        }
        false
    }

    /// Quark confinement: combine quarks into hadrons.
    pub fn quark_confinement(&mut self, rng: &mut impl Rng) -> Vec<Particle> {
        if self.temperature > T_QUARK_HADRON {
            return Vec::new();
        }

        let mut hadrons = Vec::new();

        // Collect indices of ups and downs
        let mut up_indices: Vec<usize> = self
            .particles
            .iter()
            .enumerate()
            .filter(|(_, p)| p.particle_type == ParticleType::Up)
            .map(|(i, _)| i)
            .collect();
        let mut down_indices: Vec<usize> = self
            .particles
            .iter()
            .enumerate()
            .filter(|(_, p)| p.particle_type == ParticleType::Down)
            .map(|(i, _)| i)
            .collect();

        let mut to_remove = Vec::new();

        // Form protons (uud)
        while up_indices.len() >= 2 && !down_indices.is_empty() {
            let u1_idx = up_indices.pop().unwrap();
            let u2_idx = up_indices.pop().unwrap();
            let d1_idx = down_indices.pop().unwrap();

            let pos = self.particles[u1_idx].position;
            let mom = [
                self.particles[u1_idx].momentum[0]
                    + self.particles[u2_idx].momentum[0]
                    + self.particles[d1_idx].momentum[0],
                self.particles[u1_idx].momentum[1]
                    + self.particles[u2_idx].momentum[1]
                    + self.particles[d1_idx].momentum[1],
                self.particles[u1_idx].momentum[2]
                    + self.particles[u2_idx].momentum[2]
                    + self.particles[d1_idx].momentum[2],
            ];

            to_remove.push(u1_idx);
            to_remove.push(u2_idx);
            to_remove.push(d1_idx);

            let id = self.alloc_id();
            let proton = Particle {
                id,
                particle_type: ParticleType::Proton,
                position: pos,
                momentum: mom,
                spin: random_spin(rng),
                color: None,
            };
            hadrons.push(proton);
        }

        // Form neutrons (udd)
        while !up_indices.is_empty() && down_indices.len() >= 2 {
            let u1_idx = up_indices.pop().unwrap();
            let d1_idx = down_indices.pop().unwrap();
            let d2_idx = down_indices.pop().unwrap();

            let pos = self.particles[u1_idx].position;
            let mom = [
                self.particles[u1_idx].momentum[0]
                    + self.particles[d1_idx].momentum[0]
                    + self.particles[d2_idx].momentum[0],
                self.particles[u1_idx].momentum[1]
                    + self.particles[d1_idx].momentum[1]
                    + self.particles[d2_idx].momentum[1],
                self.particles[u1_idx].momentum[2]
                    + self.particles[d1_idx].momentum[2]
                    + self.particles[d2_idx].momentum[2],
            ];

            to_remove.push(u1_idx);
            to_remove.push(d1_idx);
            to_remove.push(d2_idx);

            let id = self.alloc_id();
            let neutron = Particle {
                id,
                particle_type: ParticleType::Neutron,
                position: pos,
                momentum: mom,
                spin: random_spin(rng),
                color: None,
            };
            hadrons.push(neutron);
        }

        // Remove consumed quarks (sort descending to keep indices valid)
        to_remove.sort_unstable();
        to_remove.dedup();
        for &idx in to_remove.iter().rev() {
            if idx < self.particles.len() {
                self.particles.swap_remove(idx);
            }
        }

        // Add new hadrons
        for h in &hadrons {
            self.particles.push(h.clone());
        }
        self.total_created += hadrons.len() as u64;

        hadrons
    }

    /// Evolve all particles by one time step.
    pub fn evolve(&mut self, dt: f64) {
        for p in &mut self.particles {
            let m = p.particle_type.mass();
            if m > 0.0 {
                for i in 0..3 {
                    p.position[i] += p.momentum[i] / m * dt;
                }
            } else {
                let p_mag = (p.momentum[0].powi(2) + p.momentum[1].powi(2) + p.momentum[2].powi(2))
                    .sqrt()
                    .max(1e-20);
                for i in 0..3 {
                    p.position[i] += p.momentum[i] / p_mag * C * dt;
                }
            }
        }
    }

    /// Count particles by type.
    pub fn particle_counts(&self) -> Vec<(ParticleType, usize)> {
        use std::collections::HashMap;
        let mut map: HashMap<ParticleType, usize> = HashMap::new();
        for p in &self.particles {
            *map.entry(p.particle_type).or_insert(0) += 1;
        }
        let mut v: Vec<_> = map.into_iter().collect();
        v.sort_by_key(|(pt, _)| *pt as u8);
        v
    }

    /// Count of protons.
    pub fn proton_count(&self) -> usize {
        self.particles
            .iter()
            .filter(|p| p.particle_type == ParticleType::Proton)
            .count()
    }

    /// Count of neutrons.
    pub fn neutron_count(&self) -> usize {
        self.particles
            .iter()
            .filter(|p| p.particle_type == ParticleType::Neutron)
            .count()
    }

    /// Remove all protons and neutrons (after nucleosynthesis consumes them).
    pub fn remove_hadrons(&mut self) {
        self.particles.retain(|p| {
            p.particle_type != ParticleType::Proton && p.particle_type != ParticleType::Neutron
        });
    }

    /// Remove protons and electrons used in recombination.
    pub fn remove_for_recombination(&mut self, proton_count: usize, electron_count: usize) {
        let mut p_left = proton_count;
        let mut e_left = electron_count;
        self.particles.retain(|p| {
            if p_left > 0 && p.particle_type == ParticleType::Proton {
                p_left -= 1;
                return false;
            }
            if e_left > 0 && p.particle_type == ParticleType::Electron {
                e_left -= 1;
                return false;
            }
            true
        });
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn gauss(rng: &mut impl Rng, mean: f64, std: f64) -> f64 {
    // Box-Muller transform
    let u1: f64 = rng.gen::<f64>().max(1e-15);
    let u2: f64 = rng.gen::<f64>();
    mean + std * (-2.0 * u1.ln()).sqrt() * (2.0 * PI * u2).cos()
}

fn random_spin(rng: &mut impl Rng) -> Spin {
    if rng.gen::<bool>() {
        Spin::Up
    } else {
        Spin::Down
    }
}

fn random_color(rng: &mut impl Rng) -> ColorCharge {
    match rng.gen_range(0..3) {
        0 => ColorCharge::Red,
        1 => ColorCharge::Green,
        _ => ColorCharge::Blue,
    }
}
